### PR TITLE
Inline List.Clear

### PR DESCRIPTION
**Pull request by benaadams**
Even though the two paths are small post jit (non ref and ref) the combination pre-jit in the same function makes it look to big too inline.

Contributes to https://github.com/dotnet/corefx/issues/18390

/cc @jkotas @AndyAyersMS

### CODE REVIEW

**Comment by benaadams** [comment]
Asm to inline

Non reference asm
```asm
G_M4562_IG01:
       0F1F440000           nop      

G_M4562_IG02:
       33C0                 xor      eax, eax
       894118               mov      dword ptr [rcx+24], eax
       FF411C               inc      dword ptr [rcx+28]

G_M4562_IG03:
       C3                   ret    
```
Reference asm
```asm
G_M58440_IG01:
       0F1F440000           nop      

G_M58440_IG02:
       4C8B01               mov      r8, qword ptr [rcx]
       448B4118             mov      r8d, dword ptr [rcx+24]
       33D2                 xor      edx, edx
       895118               mov      dword ptr [rcx+24], edx
       FF411C               inc      dword ptr [rcx+28]
       4585C0               test     r8d, r8d
       7E10                 jle      SHORT G_M58440_IG04
       488B4908             mov      rcx, gword ptr [rcx+8]
       33D2                 xor      edx, edx
       488D0500000000       lea      rax, [(reloc)]

G_M58440_IG03:
       48FFE0               rex.jmp  rax

G_M58440_IG04:
       C3                   ret     
```

**Comment by benaadams** [comment]
I think the `mov      r8, qword ptr [rcx]` is extra in the ref version?

**Comment by omariom** [comment]
@AndyAyersMS
Are there plans to make jit take into account the final size of methods?

**Comment by benaadams** [comment]
@omariom this its covered by this issue? https://github.com/dotnet/coreclr/issues/10731

**Comment by omariom** [comment]
@benaadams 
I even liked it but then completely forgot  it :smile:
After that change JIT will be inlining List.Clear only for non RoCR types.

Great to see that List.Clear is now just 2 machine instructions :+1:

**Comment by AndyAyersMS** [comment]
@omariom no, not exactly -- to inline this w/o the aggressive inline attribute the jit would have to anticipate that a type optimization was going to happen in the inlinee and then estimate the impact this optimization would have on the method size. When and where the type optimization happens is not as crucial for the inliner.

When we cost an inlinee today, we do so by just looking at its IL stream. There may be things we can spot in our fairly shallow evaluation of that stream to suggest that type based optimizations are likely to have a size impact. Estimating the magnitude of the impact may be a bit trickier.

An alternative that we could pursue in crossgen is to generate code for methods "bottom-up" -- callees before callers. Then we could flag callees where final size is considerably less than would be expected from their IL size and use this as a part of the size estimates for inlines when compiling callers.

In a tiered jit world we might also be able to profit from such smaller-than-expected observations and reconsider inlining methods that we previously rejected.

**Comment by AndyAyersMS** [comment]
Would have to drill in to see why that extra mov hangs around. It could be a vestige of the runtime lookup we create initially to call `IsReference...` and so perhaps something that #10946 cleans up.

**Comment by mikedn** [comment]
@AndyAyersMS Isn't the extra mov there because it has an exception side effect? I suppose it could be removed because the following mov happens to have the same side effect but that's not something that the JIT does today.

**Comment by AndyAyersMS** [comment]
Yeah, it could be. I haven't looked at the jit dumps to see exactly what is going on.

**Comment by mikedn** [comment]
Dump looks like this:
```
[000006] ------------             *  stmtExpr  void  (IL 0x000...  ???)
[000003] #--X--------             |  /--*  indir     long
[000002] ------------             |  |  \--*  lclVar    ref    V00 this
[000005] -A-X--------             \--*  =         long
[000004] D------N----                \--*  lclVar    long   V02 tmp0
```
Once introduced it looks like this indir can't easily go away.

Come to think of it this is seemingly strange side effect of an implementation detail. You'd think that a method like
```C#
void Clear() {
    if (RuntimeHelpers.IsReferenceOrContainsReferences<T>()) {
        // do nothing
    }
}
```
should not throw null ref exception if `this` is `null` but it does that because the type context is obtained via `this`.

**Comment by AndyAyersMS** [comment]
Yeah, that's what I was getting at, but I didn't articulate it very well. Let me try again...

The runtime helper is static, so we should be able to call it with a null `this`. But our runtime lookup expansion introduces a dereference. So evidently there's some implicit (or perhaps explicit) requirement when calling a method that needs to do a runtime lookup off of `this` that callers must do a null check, so `this` is known to be not-null on entry, and hence the deref in the lookup path will never fail and and is not semantically motivated, so can be removed, despite it looking like something that can cause an exception.

**Comment by AndyAyersMS** [comment]
Ecma I.8.4.2 says that instance methods on generic types can't be called with null `this` pointers. But I don't know how this is enforced or whether trying to leverage this puts the jit into the temptingly dangerous "undefined behavior" bucket where it can just assume what it expects to be true.

Most times these methods are invoked via `callvirt` so presumably the primary enforcement is on the caller side. But that's not required. Hacking up a simple example like the above to pass a null this and use `call` opcode still triggers NRE because of the residual internal check. But presumably per I.8.4.2 we could optimize that away.

But unless we modify the jit to always remove the null check (even in minopts/debuggable mode), this would have the annoying effect of possibly changing behavior because of optimization.

We might also consider enforcing this more directly by having the jit interface demand a null check for any generic instance method call, regardless of opcode.

@jkotas any thoughts?

**Comment by jkotas** [comment]
I think it is fine to have undefined behavior wrt NREs when a method on generic type is called with null thisptr.

It is implementation dependent which exact operations do dictionary lookups that would trigger the NRE. I do not think it makes sense to try to set the exact set of the operations in stone. Also, we do not have any extensive tests to verify that the NREs are maintained in the exact same spots, so it is likely it have changed many times without us noticing.
